src/ocaml/typing/env.ml

```diff
@@ -496,6 +496,28 @@ type type_descriptions = type_descr_kind
 
 let in_signature_flag = 0x01
 
+let stamped_changes =
+  s_table Stamped_hashtable.create_changes ()
+
+let stamped_add table path value =
+  let rec path_stamp = function
+    | Pident id -> Ident.stamp id
+    | Pdot (t, _) -> path_stamp t
+    | Papply (t1, t2) -> Int.max (path_stamp t1) (path_stamp t2)
+  in
+  let stamp = path_stamp path in
+  let stamp = if stamp = 0 then None else Some stamp in
+  Stamped_hashtable.add table ?stamp path value
```

Comment:
There are two issues, they might be addressable but I didn't find how:

1) We don't just need to iterate on the keys, we also need to iterate on the hashtables, I am not sure what would be a simple way to iterate on all `fcomp_cache`/`fcomp_subst_cache` in `Persistent_env`?
This is why I introduced the `stamped_changes` indirection: it gathers the pairs `(table, key)` for all the recently added stamped keys.

2) Even if we find a way to iterate on all tables and keys, the amortized complexity will change from something like $O(\Delta n)$ to $O(n)$, which is probably something we want to avoid if we want to care about performance. The logic in `Stamped_hashtable` is careful to maximize the work that can be shared between snapshots to avoid doing redundant computation. 

If we find a better way to identify precisely the tables in Persistent_env, a simpler design might be possible.

Reply:
To summarize:
(1) unless we find a way to iterate the tables, we need a separate structure to keep track of the changes
(2) by keeping it lazily sorted, we get strong guarantees on the performance

(1) complicates the interface, which can impact maintenance.
(2) complicates the implementation, but it has to be done only once, so it's mostly free now :).